src/scrot.c

```diff
@@ -285,33 +290,51 @@ static Imlib_Image scrotGrabAutoselect(void)
     return im;
 }
 
-/* similar to sleep, but deals with EINTR.
- * ideally, we'd want to sleep against an absolute time to prevent any drift
- * but clock_nanosleep doesn't seem to be widely implemented.
- */
-static void scrotSleep(int sec)
+void scrotDoDelay(void)
 {
-    assert(sec > 0);
-    struct timespec delay = { .tv_sec = sec };
-    while (nanosleep(&delay, &delay) < 0 && errno == EINTR);
+    if (!opt.delay)
+        return;
+    if (opt.countdown) {
+        dprintf(STDERR_FILENO, "Taking shot in ");
+        for (int i = opt.delay; i > 0; i--) {
+            dprintf(STDERR_FILENO, "%d.. ", i);
+            opt.delayStart.tv_sec++;
+            scrotWaitUntil(&opt.delayStart);
+        }
+        dprintf(STDERR_FILENO, "0.\n");
+    } else {
+        opt.delayStart.tv_sec += opt.delay;
+        scrotWaitUntil(&opt.delayStart);
+    }
 }
 
-void scrotDoDelay(void)
+/* scrotWaitUntil: clock_nanosleep with a simpler interface and no EINTR nagging
+ */
+static void scrotWaitUntil(const struct timespec *time)
 {
-    if (opt.delay) {
-        if (opt.countdown) {
-            int i;
-
-            dprintf(STDERR_FILENO, "Taking shot in %d.. ", opt.delay);
-            scrotSleep(1);
-            for (i = opt.delay - 1; i > 0; i--) {
-                dprintf(STDERR_FILENO, "%d.. ", i);
-                scrotSleep(1);
-            }
-            dprintf(STDERR_FILENO, "0.\n");
-        } else
-            scrotSleep(opt.delay);
-    }
+#if defined(HAVE_CLOCK_NANOSLEEP)
+    while (clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, time, NULL)
+        == -1 && errno == EINTR);
+#else
+    /* OpenBSD and OS X lack clock_nanosleep(), so we use a trivial algorithm to
+     * correct for drift and call nanosleep().
```

Comment:
```suggestion
     * reduce drift and call nanosleep().
```

Reply:
it doesn't only reduce drift, it does correct.

Reply:
Time still passes between the `clock_gettime` call and the `nanosleep` call. If unfortunate, the kernel might perform a context switch and put our process to sleep while the timespecsub is happening, which can have measurable impact (depending on the kernel config/scheduler etc).

Reply:
That's true, but the same happens between `clock_nanosleep()` and anything after. 

By drift, I mean if you call `nanosleep()` in a loop, you have drift proportional to O(n) where n is the amount of times you call `nanosleep()`. I.e., if you sleep 1 second and do a task that takes 100ms in a loop, and you do that task 10 times, you will be 1 second off. Calling `clock_gettime()` and sleeping for the difference between the target time and the current time reduces that drift to O(1).